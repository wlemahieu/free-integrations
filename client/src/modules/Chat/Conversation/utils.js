import React from 'react';
import { each, map } from 'lodash';
import { Timeline } from 'antd';

export const createTimelineData = (inputs, responses) => {
	const timeline = [];
	each(inputs, (value, key) => {
		timeline.unshift(value);
		if (responses[key]) {
			timeline.unshift(responses[key]);
		}
	});
	return timeline;
};

export const createTimelineItems = (inputs, responses) => {
	const timeline = createTimelineData(inputs, responses);
	const timelineEven = timeline.length % 2 === 0;
	return map(timeline, (text, index) => {
		const indexEven = index % 2 === 0;
		const color = (!indexEven && timelineEven) || (indexEven && !timelineEven) ? 'blue' : 'green';
		const key = color === 'blue' ? `me-${index}` : `ai-${index}`;
		return <Timeline.Item color={color} key={key}>{text}</Timeline.Item>;
	});
};
