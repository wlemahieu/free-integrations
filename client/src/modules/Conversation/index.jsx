import React, { PureComponent } from 'react';
import { Col, Row, Timeline } from 'antd';

class Conversation extends PureComponent {
	render() {
		return (
			<Row className="centered" style={{ paddingTop: '60px' }}>
				<Col span={12} offset={6}>
					<Timeline mode="alternate">
						<Timeline.Item>My name is Wesley. How old are you CleverWatson?</Timeline.Item>
						<Timeline.Item color="green">I'm doing fine, thank you. What's your name?</Timeline.Item>
						<Timeline.Item>Hello there. How are you?</Timeline.Item>
					</Timeline>
				</Col>
			</Row>
		);
	}
}

export default Conversation;
