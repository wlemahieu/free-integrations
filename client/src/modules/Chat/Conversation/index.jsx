import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Scrollbars } from 'react-custom-scrollbars';
import { Col, Row, Timeline } from 'antd';
import { createTimelineItems } from './utils.js';

class Conversation extends PureComponent {
	render() {
		const chat = this.props.chat;
		return (
			<Row className="centered" style={{ paddingTop: '60px' }}>
				<Col span={12} offset={6}>
					<Scrollbars style={{ height: 600 }}>
						<Timeline mode="alternate" style={{ paddingRight: '20px' }}>
							{createTimelineItems(chat.inputs, chat.responses)}
						</Timeline>
					</Scrollbars>
				</Col>
			</Row>
		);
	}
}

Conversation.propTypes = {
	input: PropTypes.string,
	inputs: PropTypes.array,
	responses: PropTypes.array
};

Conversation.defaultProps = {
	input: '',
	inputs: [],
	responses: []
};

export default Conversation;
