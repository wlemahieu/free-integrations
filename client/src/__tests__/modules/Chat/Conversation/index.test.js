import React from 'react';
import { every, isArray, isString } from 'lodash';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';
import { Scrollbars } from 'react-custom-scrollbars';
import { Col, Row, Timeline } from 'antd';
import Conversation from 'src/modules/Chat/Conversation';
import { createTimelineData, createTimelineItems } from 'src/modules/Chat/Conversation/utils.js';

const props = {
	inputs: ['I feel good.', 'Hello'],
	responses: ['I am glad you feel good.', 'Hey there, how are you?']
};
const timeline = createTimelineData(props.inputs, props.responses);

let wrapper;
beforeEach(() => {
	wrapper = shallow(<Conversation {...props} />);
});

describe('<Conversation />', () => {
	it('matches snapshot', () => {
		expect(toJson(wrapper)).toMatchSnapshot();
	});
	it('renders 1 <Row> component', () => {
		expect(wrapper.find('Row')).toHaveLength(1);
	});
	it('renders 1 <Col> component', () => {
		expect(wrapper.find('Col')).toHaveLength(1);
	});
	it('renders 1 <Scrollbars> component', () => {
		expect(wrapper.find('Scrollbars')).toHaveLength(1);
	});
	it('renders 1 <Timeline> component', () => {
		expect(wrapper.find('Timeline')).toHaveLength(1);
	});
	it(`renders 4 <TimelineItem> components`, () => {
		expect(wrapper.find('TimelineItem')).toHaveLength(4);
	});
	it('renders <Row> className = \'centered\'', () => {
		expect(wrapper.find('Row').props().className).toEqual('centered');
	});
	it('renders <Timeline> mode = \'alternate\'', () => {
		expect(wrapper.find('Timeline').props().mode).toEqual('alternate');
	});
});

describe('createTimelineData()', () => {
	it('returns an array', () => {
		expect(isArray(timeline)).toBe(true);
	});
	it('renders an array of strings', () => {
		expect(every(timeline, isString)).toBe(true);
	});
});

describe('createTimelineItems()', () => {
	it('renders an array of components', () => {
		const timeline = createTimelineData(props.inputs, props.responses);
		expect(isArray(timeline)).toBe(true);
	});
});
