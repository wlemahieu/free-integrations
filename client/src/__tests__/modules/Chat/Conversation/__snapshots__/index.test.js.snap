// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Conversation /> matches snapshot 1`] = `
<Row
  className="centered"
  gutter={0}
  style={
    Object {
      "paddingTop": "60px",
    }
  }
>
  <Col
    offset={6}
    span={12}
  >
    <Scrollbars
      autoHeight={false}
      autoHeightMax={200}
      autoHeightMin={0}
      autoHide={false}
      autoHideDuration={200}
      autoHideTimeout={1000}
      hideTracksWhenNotNeeded={false}
      renderThumbHorizontal={[Function]}
      renderThumbVertical={[Function]}
      renderTrackHorizontal={[Function]}
      renderTrackVertical={[Function]}
      renderView={[Function]}
      style={
        Object {
          "height": 600,
        }
      }
      tagName="div"
      thumbMinSize={30}
      universal={false}
    >
      <Timeline
        mode="alternate"
        reverse={false}
        style={
          Object {
            "paddingRight": "20px",
          }
        }
      >
        <TimelineItem
          color="green"
          key="ai-0"
          pending={false}
          position=""
        >
          Hey there, how are you?
        </TimelineItem>
        <TimelineItem
          color="blue"
          key="me-1"
          pending={false}
          position=""
        >
          Hello
        </TimelineItem>
        <TimelineItem
          color="green"
          key="ai-2"
          pending={false}
          position=""
        >
          I am glad you feel good.
        </TimelineItem>
        <TimelineItem
          color="blue"
          key="me-3"
          pending={false}
          position=""
        >
          I feel good.
        </TimelineItem>
      </Timeline>
    </Scrollbars>
  </Col>
</Row>
`;
